### PR TITLE
PLT-6359 Create SQL index tables only once fully synced

### PR DESCRIPTION
**Pull request by koslambrou**
Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense and have useful messages
    - [ ] Important changes are reflected in changelog.d of the affected packages
    - [ ] Relevant tickets are mentioned in commit messages
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Targeting main unless this is a cherry-pick backport
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] If relevant, reference the ADR in the PR and reference the PR in the ADR
    - [ ] Reviewer requested

### CODE REVIEW

**Comment by berewt** [comment]
I like the general idea but I'm not sure I'm comfortable using `Tracer` to add the index if we have another mean. If possible, moving to an action (or a list of actions) to trigger once we're synced would be my preference, but I'm open to discussing both solutions.

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers.hs [hunk 1]
```diff
@@ -69,13 +71,13 @@ buildIndexers securityParam catchupConfig utxoConfig mintEventConfig epochStateC
     epochStateBuilder securityParam catchupConfig epochStateConfig mainLogger path
 
   StandardWorker utxoMVar utxoWorker <-
-    utxoBuilder securityParam catchupConfig utxoConfig txBodyCoordinatorLogger path
+    utxoBuilder securityParam catchupConfig utxoConfig txBodyCoordinatorLogger textLogger path
```
Comment (koslambrou) [inline marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers.hs:74]: Don't like that part of passing two different loggers here. Suggestions accepted.

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/Spent.hs [hunk 2]
```diff
@@ -133,6 +133,50 @@ mkSpentIndexer path = do
...
+                  WHERE type='index' AND name='|]
+                  <> fromString indexName
+                  <> "'"
+              )
+      unless doesIndexExist $ do
+        logInfo stdoutTrace $
+          "Creating SQL index table '" <> Text.pack indexName <> "' for Spent indexer..."
+        liftIO $ SQL.execute_ c createIndexStatement
+        logInfo stdoutTrace $
+          "Finished creating SQL index table '" <> Text.pack indexName <> "' for Spent indexer..."
```
Comment (koslambrou) [inline marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/Spent.hs:178]: This is slightly duplicated across some indexers, but I don't see the ideal place to put it.

Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/Spent.hs:178]: Add a helper function to either `Core.Marconi.Indexer.SQLite` or `Core.Marconi.Transformer.WithCatchup`?

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/MintTokenEvent.hs [hunk 3]
```diff
@@ -354,15 +354,51 @@ mkMintTokenIndexer dbPath = do
...
+                  <> fromString indexName
+                  <> "'"
+              )
+      unless doesIndexExist $ do
+        logInfo stdoutTrace $
+          "Creating SQL index table '" <> Text.pack indexName <> "' for MintTokenEvent indexer..."
+        liftIO $ SQL.execute_ c createIndexStatement
+        logInfo stdoutTrace $
+          "Finished creating SQL index table '" <> Text.pack indexName <> "' for MintTokenEvent indexer..."
+
```
Comment (berewt) [inline marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/MintTokenEvent.hs:401]: As `CREATE INDEX IF NOT EXISTS` already takes care of not doing anything if we have the index in place, do we want to check it manually to provide a feedback?

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/Utxo.hs [hunk 4]
```diff
@@ -158,6 +159,42 @@ mkUtxoIndexer path = do
...
+            <> " ON utxo (address)"
+    createIndexTable c addressIndexName createAddressIndexStatement
+
+    let slotNoIndexName = "utxo_slotNo"
+        createSlotNoIndexStatement =
+          "CREATE INDEX IF NOT EXISTS "
+            <> fromString slotNoIndexName
+            <> " ON utxo (slotNo)"
+    createIndexTable c slotNoIndexName createSlotNoIndexStatement
+  where
```
Comment (berewt) [inline marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/Utxo.hs:178]: Minor: I really think we should avoid mixing `let` and `where`.

**Review by berewt — APPROVED** [review]
It looks good to me, should we handle the primary kes this way too? In this ticket or in another one.

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/MintTokenEvent.hs (hunk 3 above)
Comment (koslambrou) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/MintTokenEvent.hs:401]: Exactly, it's just for the logging.

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/Spent.hs (hunk 2 above)
Comment (koslambrou) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/Spent.hs:178]: I guess I can put it in `Core.Marconi.Indexer.SQLite`, but feels a bit out-of-place.

**Comment by koslambrou** [comment]
> It looks good to me, should we handle the primary kes this way too? In this ticket or in another one.

The issue with the primary key is that it also checks for uniqueness, which (I guess) you want to fail as soon as possible if the constraint is violated?

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/Spent.hs (hunk 2 above)
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/Spent.hs:178]: In a submodule of one of these two?

**Comment by berewt** [comment]
> > It looks good to me, should we handle the primary kes this way too? In this ticket or in another one.
> 
> The issue with the primary key is that it also checks for uniqueness, which (I guess) you want to fail as soon as possible if the constraint is violated?

Yes, at the same time if we're able to sync with the primary key available during a test on mainnet, there's no reason to hit a failure on a next run. We could handle it with a flag maybe?

Although, we can keep it as is in this PR and think about how to deal with this specific case in another ticket.

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/Spent.hs (hunk 2 above)
Comment (koslambrou) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/Spent.hs:178]: I put it in `Marconi.Core.Transformer.WithCatchup.SQLite`.

**Comment by koslambrou** [comment]
> Yes, at the same time if we're able to sync with the primary key available during a test on mainnet, there's no reason to hit a failure on a next run. We could handle it with a flag maybe?

Good point! I'll create a ticket for that.

**Pull request merged**